experimental/ir/fdp.go

```diff
@@ -0,0 +1,443 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ir
+
+import (
+	"strings"
+
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/descriptorpb"
+
+	"github.com/bufbuild/protocompile/experimental/ast/predeclared"
+	"github.com/bufbuild/protocompile/experimental/ast/syntax"
+	"github.com/bufbuild/protocompile/experimental/ir/presence"
+	"github.com/bufbuild/protocompile/experimental/seq"
+	"github.com/bufbuild/protocompile/internal/ext/slicesx"
+	"github.com/bufbuild/protocompile/internal/intern"
+	"github.com/bufbuild/protocompile/wellknownimports"
+)
+
+// FDS code-generates a FileDescriptorSet for the given files, and returns the
+// result as an encoded byte slice.
+//
+// The resulting FDS is always fully linked: it contains all dependencies except
+// the WKTs, and all names are fully-qualified.
+func FDS(files []File, options ...CodegenOption) ([]byte, error) {
+	var c codegen
+	for _, opt := range options {
+		if opt != nil {
+			opt(&c)
+		}
+	}
+
+	fds := new(descriptorpb.FileDescriptorSet)
+	c.files(files, fds)
+	return proto.Marshal(fds)
+}
+
+// FDP code-generates a single FileDescriptorProto for file, and returns the
+// result as an encoded byte slice.
+//
+// The resulting FDP is fully linked: all names are fully-qualified.
+func FDP(file File, options ...CodegenOption) ([]byte, error) {
+	var c codegen
+	for _, opt := range options {
+		if opt != nil {
+			opt(&c)
+		}
+	}
+
+	fdp := new(descriptorpb.FileDescriptorProto)
+	c.file(file, fdp)
+	return proto.Marshal(fdp)
+}
+
+// CodegenOption is an option to pass to [DescriptorSet].
+type CodegenOption func(*codegen)
+
+// IncludeWKTs is an option for including FDPs for the Well-Known Types in
+// codegen output.
+func IncludeWKTs() CodegenOption {
+	return func(c *codegen) { c.includeWKTs = true }
+}
+
+// IncludeSourceCodeInfo is an option that requires generating source code info
+// for the descriptors.
+func IncludeSourceCodeInfo() CodegenOption {
+	return func(c *codegen) { c.includeSourceCodeInfo = true }
+}
+
+type codegen struct {
+	includeWKTs           bool
+	includeSourceCodeInfo bool
+
+	currentFile File
+}
+
+func (c *codegen) files(files []File, fds *descriptorpb.FileDescriptorSet) {
+	if c.includeSourceCodeInfo {
+		panic("protocompile/ir: NYI: IncludeSourceCodeInfo()")
+	}
+
+	// Build up all of the imported files. We can't just pull out the transitive
+	// imports for each file because we want the result to be sorted
+	// topologically.
+	for file := range TopoSort(files...) {
+		if !c.includeWKTs && wellknownimports.StandardImport(file.Path()) != "" {
+			continue
+		}
```

Comment:
This does not seem like a great way to detect it being a well-known import. I would instead expect that we somehow traced the source contents to the `wellknownimports` package. I definitely don't see any value in trying to prune these files if this is the heuristic used to prune them.